Fixes for the problems reported by Alexander for session 2:

ControlX opened the pull request:
- Moved jest, html-webpack-plugin and webpack-dev-server to devDependencies
- Commiting .gitignore which got left out.
- Fixed mode options in package.json as they are overriding webpackconfigs.

shuspes commented on package.json:
```diff
@@ -24,21 +24,24 @@
+    "redux": "^4.0.5"
   },
   "devDependencies": {
     "@babel/core": "^7.11.1",
```
You do not have **style-loader** package in dev dependencies. That's why webpack has issues with css files

shuspes commented on package.json:
```diff
@@ -24,21 +24,24 @@
   },
   "devDependencies": {
     "@babel/core": "^7.11.1",
     "@babel/preset-env": "^7.11.0",
```
Your project should have package-lock.json file

shuspes commented on src/index.js:
```diff
@@ -1,14 +1,11 @@
-        <h1>Webpack and other configurations in ReactJS.</h1>
-        <h2>Application is running on the environment: {process.env.NODE_ENV}</h2>
+    <div className='main-div-properties'>
+        {<AppComponent />}
```
You do not need to wrap AppComponent in curly braces

shuspes commented on src/Components/Component1/Footer.jsx:
```diff
@@ -0,0 +1,7 @@
+
+export default function Footer(){
+    return (
+        {}//Nothing for now
```
To render nothing you should use **null**. When you return empty object, browser fails with error

shuspes commented on src/Components/Component1/Constants.js:
```diff
@@ -0,0 +1,86 @@
+export default DUMMY_JSON_PAYLOAD = [
```
When you ```export default``` something, you should not name it somehow - [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export)

shuspes commented on src/Components/Component1/MoviesList.jsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import MovieCard from './MovieCard'
+import { DUMMY_JSON_PAYLOAD } from './Constants'
```
default export do not need curly braces

shuspes commented on src/Components/Component1/MoviesList.jsx:
```diff
@@ -0,0 +1,21 @@
+    return (
+        <>
+        {movies.map((movie) => (
+            <div className='gallery'>
```
every element in React list should have **key** - [documentation](https://reactjs.org/docs/lists-and-keys.html#basic-list-component)

shuspes commented on webpack.config.dev.js:
```diff
@@ -5,7 +5,7 @@ module.exports = {
+       path: path.join(__dirname, '/src/index.js')},
     module:{
        rules:[{
           test: /\.js$/,
```
Update test rule with this value - ```/\.(js|jsx)$/```

shuspes commented on webpack.config.dev.js:
```diff
@@ -16,8 +16,20 @@ module.exports = {
     },
+    resolve: {
+      modules: [path.resolve(__dirname, "/src"), "node_modules"],
+      extensions: [".js", ".json"],
```
Please, add ```".jsx"``` to the extensions

shuspes commented on webpack.config.dev.js:
```diff
@@ -5,7 +5,7 @@ module.exports = {
     output: {
        filename: 'build.js',
-       path: path.join(__dirname, '/dev/dist')},
+       path: path.join(__dirname, '/src/index.js')},
```
You have incorrect output path

shuspes reviewed (changes requested):
Your issue with build proccess was due to the fact that you do not build jsx files and do not resolve modules for them. Also, please, fix path for the image file.

ControlX commented:
Hi @shuspes ,
 
- I have fixed all the places you have mentioned.
- App is working now
- Refactored the code and placed files in their respective folders

Kindly review.

Thanks.

shuspes commented on src/Components/Component1/utils/Constants.js:
```diff
@@ -0,0 +1,88 @@
+export default function DummyInvisionListing() {
```
if you want to use function, please, name it as an action

shuspes commented on src/index.js:
```diff
@@ -1,11 +1,11 @@
 import ReactDOM from 'react-dom';
 import AppComponent from './Components/Component1/AppComponent'
-import './Components/Component1/styles.css'
+import './Components/Component1/css/styles.css'
```
This styles should be inserted in the proper component: Components/Component1

shuspes replied:
Also, you can remove extra **Component1** folder, you don't need it

shuspes commented on src/Components/Component1/MoviesList.jsx:
```diff
@@ -1,17 +1,18 @@
-export default function MoviesList(){
+
+export default function MoviesList(props){
+    let movies = props.listing;
```
You do not need this extra variable, please, use **listing** props directly

shuspes commented on src/Components/Component1/MoviesList.jsx:
```diff
@@ -1,17 +1,18 @@
+                year= {movie.year}
                 description={movie.description}
             />
             </div>
```
Where is propTypes for this component?

shuspes commented on src/Components/Component1/MovieCard.jsx:
```diff
@@ -1,13 +1,20 @@
+            </div>
         </>    
     )
 }
```
Not full list of props in the propTypes

shuspes commented on src/Components/Component1/ErrorBoundary.jsx:
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+
+function ErrorBoundary(props){
```
This ErrorBoundary saver does not work. Please, look into [documentation](https://reactjs.org/docs/error-boundaries.html) to resolve the issue. To check this component use MovieListing variable as a string in the AppComponent component.

shuspes commented on webpack.config.prod.js:
```diff
@@ -1,23 +1,35 @@
     output: {
        filename: 'build.js',
-       path: path.join(__dirname, '/prod/dist')},
+       path: path.join(__dirname, '/dev/dist')},
```
It is strange a little to use the same folder **dev** for the dev and production apps. Please, return **/prod/dist** for the prod build configuration

shuspes commented on webpack.config.prod.js:
```diff
@@ -1,23 +1,35 @@
+       },
+       {
+         test: /\.(jpe?g|png|gif|svg)$/i, 
+         loader: 'url-loader'
```
your builded app can not find images. Please, look again to the documentation how to copy images to the build folder: how to copy images to the dist directory - https://webpack.js.org/guides/asset-management/#loading-images

ControlX replied:
I didn't get this point. Please help me with what I'm doing wrong here in our next Q&A session.
When I"m running my application it is loading images at my side.

shuspes replied:
My comment was about situation when I run ```npm run build:dev``` command, go to the dist folder and open **index.html** file. In this case the app cannot find images, but I expect that dist folder after build has all files that I need for the correct app work

shuspes reviewed (approved):
Please, let me know if you need more details about any of the comments.
Your app does not have all needed controls such as search input, filter and sort selects, add movie button. Please add all needed elements according to the template to the page.

ControlX commented:
Hello @shuspes ,

I have pushed my changes for component 2.
Please review. Also, I have addressed the issues reported earlier.

Thank you.

![Screenshot 2020-09-06 at 5 16 22 AM](https://user-images.githubusercontent.com/19390507/92315225-3b132c80-f000-11ea-94e9-08ebb6fe22f8.png)

shuspes commented on src/components/MoviesList/MoviesList.jsx:
```diff
@@ -0,0 +1,31 @@
+}
+
+MoviesList.propTypes = {
+    listing: PropTypes.instanceOf('Array').isRequired
```
It is not correct way how to define type of property - [documentation](https://reactjs.org/docs/typechecking-with-proptypes.html)

ControlX replied:
Fixed.

shuspes commented on src/components/ErrorBoundary/ErrorBoundary.jsx:
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+
+function ErrorBoundary(props){
```
You should specify specific lifecycle methods to create an ErrorBoundary component - [documentation](https://reactjs.org/docs/error-boundaries.html#introducing-error-boundaries)

ControlX replied:
Need to discuss this in next Q&A. For functional components can't implement lifecycle methods...

shuspes commented on src/utils/RestUtils.js:
```diff
@@ -0,0 +1,95 @@
+            else{
+                response.data = null;
+                response.status = 400;
+                reject(Error(response))
```
Error object cannot get an object, only string - [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error)

ControlX replied:
Fixed, included error reason string.

shuspes commented on src/utils/RestUtils.js:
```diff
@@ -0,0 +1,95 @@
+                response.status = 400;
+                reject(Error(response))
+            }
+            return;
```
you do not need this **return**

ControlX replied:
Removed

shuspes commented on src/components/RouletteMain/RouletteMain.jsx:
```diff
@@ -0,0 +1,151 @@
+import SortListing from '../SortListing'
+import "regenerator-runtime"
+const MoviesList = React.lazy(() => {
+    return new Promise(resolve => setTimeout(resolve, 500)).then(() => import("../MoviesList"));
```
I understand why you add this delay before loading component, but, do not forget remove test constructions before go to production/creat a PR

ControlX replied:
Removed

shuspes commented on src/components/RouletteMain/RouletteMain.jsx:
```diff
@@ -0,0 +1,151 @@
+                <div className='parent-header-properties'>
+                    <Header
+                        isAddModalVisible={this.state.isAddModalVisible}
+                        onAddAction={this.onAddAction.bind(this)}
```
Besides, to remove **bind**, you can define handlers as an arrow functions

shuspes commented on src/components/AddMovie/AddMovie.jsx:
```diff
@@ -0,0 +1,74 @@
+    addMovieJson.runtime = e.target.value;
+  }
+
+  function resetFields(){
```
You can define this function as an arrow function and use in **onClick** only function name without strange construction with arrow function

ControlX replied:
I didn't get this. Will check in next Q&A

shuspes commented on src/components/AddMovie/AddMovie.jsx:
```diff
@@ -0,0 +1,74 @@
+import React from 'react'
+import "./add_movie.css"
+
+export default function AddMovie(props) {
```
Please, create one component for displaying **AddMovie** and **EditMovie** modals. These components display the same list of properties and have the same style and functionality.

ControlX replied:
Fixed.

shuspes commented on src/components/RouletteMain/RouletteMain.jsx:
```diff
@@ -0,0 +1,151 @@
+
+    onConfirmAction(item) {
+        let stateList = this.state.movieList;
+        stateList.forEach((obj, i, jsonObj) => { if (obj.id === item.id) { jsonObj.splice(i, 1); } });
```
Array.filter function is more apropriate in this case

ControlX replied:
Fixed

shuspes commented on src/components/RouletteMain/RouletteMain.jsx:
```diff
@@ -0,0 +1,151 @@
+
+    onEditAction(id) {
+        let stateList = this.state.movieList;
+        stateList.forEach((obj, i) => { if (obj.id === id) { this.setState({ processEditMovieField: obj }) } });
```
Please, use Array.find in this case and invoke **setState** once

ControlX replied:
Fixed.

shuspes commented on src/components/RouletteMain/RouletteMain.jsx:
```diff
@@ -0,0 +1,151 @@
+        let stateList = this.state.movieList;
+        let uniqueId = stateList.length + 1;
+        item.id = uniqueId;
+        stateList.push(item);
```
Avoid state mutations, instead of this, please, use spread syntax: ```this.setState({ movieList: [...this.state.movieList, item] })```. For example, if you extend component from the React.PureComponent, remove **this.onCloseAction();** string, list of movie will not be updated because **stateList** and **this.state.movieList** are same.

ControlX replied:
Fixed. Pushing and merging changes to master.

ControlX commented:
Merging this branch with Master.
